Iaviewer command to visualize iavl database from leveldb

ethanfrey opened the pull request:
Our nodes fell out of consensus due to inconsistent app hashes, but after locating the transaction that caused it and finding no obvious issues in the execution path, we had no real tools to figure out what the problem was.

I add a new command `cmd/iaviewer` that can also be installed via `make install` to allow forensic analysis and comparison of leveldb-backed iavl trees

We also allow an optional third argument of a version number to specify an older version of the state to use, rather than the latest one. This is interesting if you want to see what things looked like right _before_ it all went wrong.

Done:

`iaviewer data <path to db> [version]`

Prints out version, hash, size as well as every key along with a hash of the data (to keep sizes small). We try to convert as much of the key to ascii if possible, in case the app is using human-readble strings. The data is hashed, as the only thing we cared about at this point was if one key had a different value on two different nodes.

`iaviewer shape <path to db> [version]`

It is possible for two trees to have a different hash even with the same data if the insert order is different. This attempts to visualize the tree structure to locate the difference.

`iaviewer versions <path to db>`

Lists all available versions in the db, so we can compare snapshots from different points in time.

ethanfrey commented:
@Liamsi Now that cosmos mainnet is live and the launch party is over, maybe you have time to review this?

I'm happy to make changes as needed, but some feedback would be nice

Liamsi commented:
Will do tomorrow!

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+		var err error
+		version, err = strconv.Atoi(args[2])
+		if err != nil {
+			fmt.Printf("Invalid version number: %s\n", err)
```
Same as above: this should probably go to `os.Stderr`.

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+		version, err = strconv.Atoi(args[2])
+		if err != nil {
+			fmt.Printf("Invalid version number: %s\n", err)
+			os.Exit(3)
```
Nit: Besides indicating an error, is it useful to exit with a different non-zero code per failure? Here 2, below 3? 
I would suggest to exit with `os.Exit(1)` in all failure cases.

ethanfrey replied:
Okay, too verbose I guess.
I will normalize to os.Exit(1) for everything

Liamsi commented on mutable_tree.go:
```diff
@@ -45,6 +46,18 @@ func (tree *MutableTree) VersionExists(version int64) bool {
 }
 
+// AvailableVersions returns all available versions
+func (tree *MutableTree) AvailableVersions() []int {
```
It would be great to have a little test for `AvailableVersions`. I can add it it you do not have time.

ethanfrey replied:
I will look to do it a well

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+func PrintShape(tree *iavl.MutableTree) {
+	// shape := tree.RenderShape("  ", nil)
```
Delete commented out code. 

Oh wait, shouldn't this be the default and `nodeEncoder` below makes some assumptions about the key?

ethanfrey replied:
I will make a flag here

melekes replied:
Looks like this still needs to be done. Otherwise it assumes weave key format, right?

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+
+func PrintDbStats(db dbm.DB) {
+	// stats, _ := json.MarshalIndent(db.Stats(), "", "  ")
+	// fmt.Println(string(stats))
```
deleted commented out code

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		return nil, err
+	}
+	// PrintDbStats(db)
```
delete

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+	} else if strings.HasSuffix(dir, ".db/") {
+		dir = dir[:len(dir)-4]
+	} else {
+		return nil, fmt.Errorf("Database directory must end with .db")
```
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+	// TODO: doesn't work on windows!
+	cut := strings.LastIndex(dir, "/")
+	if cut == -1 {
+		return nil, fmt.Errorf("Cannot cut paths on %s", dir)
```
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Liamsi commented on immutable_tree.go:
```diff
@@ -38,6 +38,50 @@ func (t *ImmutableTree) String() string {
+	return t.renderNode(t.root, indent, 0, encoder)
+}
+
+// defaultNodeEncoder takes an id (hash, or key for leaf notes) and whether or not this is a leaf node
```
s/notes/nodes

Liamsi commented on immutable_tree.go:
```diff
@@ -38,6 +38,50 @@ func (t *ImmutableTree) String() string {
 
+// RenderShape provides a nested tree shape, ident is prepended in each level
+// Returns an array of strings, one per line, to join with "\n" or display otherwise
+func (t *ImmutableTree) RenderShape(indent string, encoder func([]byte, int, bool) string) []string {
```
At least add a comment about the `encoder`. Alternatively, introduce a commented type alias. Something along the lines of: `type NodeEncoder func(id []byte, depth int, isLeaf bool) string` or an interface:
```go
type NodeEncoder interface {
	Encode(id []byte, depth int, isLeaf bool) string
}
```

ethanfrey replied:
Will do.

Liamsi commented on mutable_tree.go:
```diff
@@ -45,6 +46,18 @@ func (tree *MutableTree) VersionExists(version int64) bool {
+			res = append(res, int(i))
+		}
+	}
+	sort.Sort(sort.IntSlice(res))
```
Shouldn't one sort be enough here?

ethanfrey replied:
There is only one sort.
`sort.IntSlice` is just a [type to enable sorting []int](https://godoc.org/sort#IntSlice)
`sort.Sort` actually does the work.

Liamsi replied:
Thanks for the clarification!

Liamsi commented on mutable_tree.go:
```diff
@@ -45,6 +46,18 @@ func (tree *MutableTree) VersionExists(version int64) bool {
 	return tree.versions[version]
 }
 
+// AvailableVersions returns all available versions
```
Nit: could add that the versions are sorted.

Liamsi commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+func parseWeaveKey(key []byte) string {
+	cut := bytes.IndexRune(key, ':')
+	if cut == -1 {
+		return encodeId(key)
```
OK. I see the fallback here works even if the key is not in that particular format.

ethanfrey replied:
Do you have a specific ascii prefix format for cosmos-sdk keys?
I am happy to add that, and maybe a cli flag to select decoder.
I just know weave, where there is readable string followed by (mostly) binary data

ethanfrey commented on cmd/iaviewer/README.md:
````diff
@@ -0,0 +1,136 @@
+```
+
+Hmmm... everything is the same, except the hash. Odd...
+So odd that I [wrote an article about it](https://medium.com/@ethan.frey)
````
Yup, I have a draft. Meant to publish yesterday. Will update this.

ethanfrey commented on cmd/iaviewer/README.md:
````diff
@@ -0,0 +1,136 @@
+make install
+```
+
+However, as this PR lives in another repo on another branch, it is a bit trickier to build.
````
Nice.

ethanfrey commented on cmd/iaviewer/main.go:
```diff
@@ -0,0 +1,164 @@
+func main() {
+	args := os.Args[1:]
+	if len(args) < 2 || (args[0] != "data" && args[0] != "shape" && args[0] != "versions") {
+		fmt.Println("Usage: iaviewer <data|shape|versions> <leveldb dir> [version number]")
```
Sure

ethanfrey commented:
Thanks for the feedback. I will likely get to this Wednesday

ethanfrey commented:
Made most requested cleanups.
Missing feature is a flag to set how we display the nodes.

I can use `nil` for only hex-display.
And add a flag `--weave` to show the weave-specific prefixes.
I'd be happy to add an `--sdk` flag to show cosmos-sdk specific readable text.

Can you point me to how the data stored in substores looks like? Just a few example keys (from root), not the whole code.

ethanfrey commented:
@Liamsi can you review again to let me know if this addresses all your comments.

Also please let me know if current cli format is good enough or if you want some application specific flags to control parsing/display. And ideally a suggestion as to what you want to see.

Thanks.

Liamsi commented:
I think hex-display is just fine. The more we can avoid adding some project specific code here the better (iavl shouldn't now about how weave encodes things, or it even shouldn't know how the SDK does it).

melekes commented:
```
nodedb.go:333:12: ndb.batch.Close undefined (type db.Batch has no field or method Close) (typecheck)
	ndb.batch.Close()
```

circle is failing with ^ but seems not related.

ethanfrey commented:
Looks related to https://github.com/tendermint/iavl/commit/cd1feb8afff203773e848d256184b9e97af18b90

I think the dependency versions changed since I pushed this. I can try to rebase and clean it up... I kinda gave up on anything moving here

melekes commented:
> I kinda gave up on anything moving here

Sorry it took us so long to merge.

>  can try to rebase and clean it up

That would be great 👍

ethanfrey commented:
Oof, merged already. Did this break develop?

Ping me here if I should clean it up

melekes commented:
> Did this break develop?

I guess

> Ping me here if I should clean it up

only if you want to & have time

Liamsi commented:
Fixed here: https://github.com/tendermint/iavl/pull/146

ethanfrey commented:
Thanks for that @Liamsi
